cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataMigratorTest.java

```diff
@@ -0,0 +1,372 @@
+    Assert.assertEquals("avalue5", v2System.getProperties(view1.toMetadataEntity()).get("akey5"));
+    Assert.assertEquals("avalue6", v2System.getProperties(artifact1.toMetadataEntity()).get("akey6"));
+
+    Assert.assertEquals("avalue11", v2Business.getProperties(app1.toMetadataEntity()).get("akey1"));
+    Assert.assertEquals("avalue2", v2Business.getProperties(flow1.toMetadataEntity()).get("akey2"));
+    Assert.assertEquals("avalue3", v2Business.getProperties(dataset1.toMetadataEntity()).get("akey3"));
+    Assert.assertEquals("avalue4", v2Business.getProperties(stream1.toMetadataEntity()).get("akey4"));
+    Assert.assertEquals("avalue5", v2Business.getProperties(view1.toMetadataEntity()).get("akey5"));
+    Assert.assertEquals("avalue6", v2Business.getProperties(artifact1.toMetadataEntity()).get("akey6"));
+  }
+
+  private void assertHistory(MetadataDataset v2System, MetadataDataset v2Business, long sTs, long bTs) {
+    verifyhistory(v2System, app1.toMetadataEntity(), sTs);
+    verifyhistory(v2System, flow1.toMetadataEntity(), sTs);
+    verifyhistory(v2System, dataset1.toMetadataEntity(), sTs);
+    verifyhistory(v2System, stream1.toMetadataEntity(), sTs);
+    verifyhistory(v2System, view1.toMetadataEntity(), sTs);
+    verifyhistory(v2System, artifact1.toMetadataEntity(), sTs);
+
+    verifyhistory(v2Business, app1.toMetadataEntity(), bTs);
+    verifyhistory(v2Business, flow1.toMetadataEntity(), bTs);
+    verifyhistory(v2Business, dataset1.toMetadataEntity(), bTs);
+    verifyhistory(v2Business, stream1.toMetadataEntity(), bTs);
+    verifyhistory(v2Business, view1.toMetadataEntity(), bTs);
+    verifyhistory(v2Business, artifact1.toMetadataEntity(), bTs);
+  }
+
+  private void assertIndex(MetadataDataset v2System) throws Exception {
+    SearchRequest sr = new SearchRequest(new NamespaceId("ns1"), "avalue1",
+                                              ImmutableSet.of(EntityTypeSimpleName.ALL), SortInfo.DEFAULT, 0,
+                                              Integer.MAX_VALUE, 1, null, false, EnumSet.of(EntityScope.USER));
+
+
+    List<MetadataEntry> entries = v2System.search(sr).getResults();
+
+    for (MetadataEntry entry : entries) {
+      Assert.assertEquals("avalue1", entry.getValue());
+    }
+  }
+
+  private void verifyhistory(MetadataDataset v2, MetadataEntity entity, long timestamp) {
+    for (Metadata metadata : v2.getSnapshotBeforeTime(ImmutableSet.of(entity), timestamp)) {
+      Assert.assertEquals(1, metadata.getProperties().size());
+    }
+  }
+
+  private long generateMetadata(DatasetId datasetId) throws Exception {
+    // Set some properties
+    write(datasetId, app1, "akey1", "avalue1");
+    write(datasetId, flow1, "akey2", "avalue2");
+    write(datasetId, dataset1, "akey3", "avalue3");
+    write(datasetId, stream1, "akey4", "avalue4");
+    write(datasetId, view1, "akey5", "avalue5");
+    write(datasetId, artifact1, "akey6", "avalue6");
+    return write(datasetId, app1, "akey1", "avalue11");
```

Comment:
You write history column with different time stamp and the get it back to verify with the timestamp written by the last write. 

It works because when you query with time in future it gives you last snapshot before that and if you are testing that then these don't even need to return the timestamp store in the history you can just query with the currentTime

Reply:
This is more precise test where we are querying the history from the last time we wrote,  not the current time. This is because we want migrated history timestamp to be same timestamp and not the current time.

Reply:
If you want to check the history from last time you wrote then don't ignore the timestamp for other write calls. History of one entity does not affect another.